benchexec/resources.py

```diff
@@ -15,191 +15,199 @@
-            cores_of_unit = cores_of_memory_region
-            logging.debug("Using memory regions as the basis for cpu core division")
-        else:
-            cores_of_unit = cores_of_package
-            logging.debug("Using physical packages as the basis for cpu core division")
+        # read & prepare mapping of cores to L3 cache
+        cores_of_L3cache = get_L3cache_mapping(allCpus_list)
+        hierarchy_levels.append(cores_of_L3cache)
+
+        # read & prepare mapping of cores to NUMA region
+        cores_of_NUMA_Region = get_NUMA_mapping(allCpus_list)
+        if cores_of_NUMA_Region:
+            hierarchy_levels.append(cores_of_NUMA_Region)
+
+        # read & prepare mapping of cores to group
+        if cores_of_NUMA_Region:
+            cores_of_group = get_group_mapping(cores_of_NUMA_Region)
+            if cores_of_group:
+                hierarchy_levels.append(cores_of_group)
+
+        # read & prepare mapping of cores to physical package
+        cores_of_package = get_package_mapping(allCpus_list)
+        hierarchy_levels.append(cores_of_package)
+
+        # read & prepare mapping of cores to die
+        cores_of_die = get_die_mapping(allCpus_list)
+        if cores_of_die:
+            hierarchy_levels.append(cores_of_die)
+
+        # read & prepare mapping of cores to cluster
+        cores_of_cluster = get_cluster_mapping(allCpus_list)
+        if cores_of_cluster:
+            hierarchy_levels.append(cores_of_cluster)
+
+        # read & prepare mapping of cores to drawer
+        cores_of_drawer = get_drawer_mapping(allCpus_list)
+        if cores_of_drawer:
+            hierarchy_levels.append(cores_of_drawer)
+
+        # read & prepare mapping of cores to book
+        cores_of_book = get_book_mapping(allCpus_list)
+        if cores_of_book:
+            hierarchy_levels.append(cores_of_book)
 
-        # read hyper-threading information (sibling cores sharing the same physical core)
-        siblings_of_core = {}
-        for core in allCpus:
-            siblings = util.parse_int_list(
-                util.read_file(
-                    f"/sys/devices/system/cpu/cpu{core}/topology/thread_siblings_list"
-                )
-            )
-            siblings_of_core[core] = siblings
-        logging.debug("Siblings of cores are %s.", siblings_of_core)
     except ValueError as e:
         sys.exit(f"Could not read CPU information from kernel: {e}")
-    return _get_cpu_cores_per_run0(
+
+    def compare_hierarchy_by_dict_length(level: HierarchyLevel):
+        """comparator function for number of elements in a dict's value list"""
+        return len(next(iter(level.values())))
+
+    hierarchy_levels.sort(key=compare_hierarchy_by_dict_length, reverse=False)
+    """sort hierarchy_levels (list of dicts) according to the dicts' value sizes"""
+
+    # add siblings_of_core at the beginning of the list to ensure the correct index
+    hierarchy_levels.insert(0, siblings_of_core)
```

Comment:
I am wondering what the problem here could be? What would need to happen such that siblings land in the wrong place? And in any case like this (e.g., if there are siblings that have different L3 caches), wouldn't the allocation be meaningless anyway?

I guess it would be better to actually check for this instead of blindly inserting `siblings_of_core` in the first position. One could for example insert `siblings_of_core` normally, then sort, and then check that it is in the expected position.

Reply:
A theoretic case where the sibling could land in the "wrong" place would be a topology without hyperthreading siblings and without shared L3 caches (each core has its own). Come to think about it, then it probably wouldnt make a difference.  
We now check whether there are identical levels and delete one of them anyway. 
The insert follows the thought of "better safe than sorry" and makes it easy to understand that hierarchy_levels[0] is the HT siblings.

Reply:
Implemented by 2f3f9640ee20cce5a124e2903accd71f51b1238b.